### PR TITLE
add ability to define annotation in other bundles

### PR DESCRIPTION
**Pull request by wodka**
Attempt to implement https://github.com/schmittjoh/JMSDiExtraBundle/issues/210

### CODE REVIEW

**Comment by Ener-Getick** [comment]
I don't like this solution, it means that your bundle has to be configured before JMSDiExtraBundle
 which is not the case usually.

What about using the bundle configuration instead (``DependencyInjection\Configuration``)

**Comment by wodka** [comment]
:/ I think it doesn't have to be defined before.

actually only once the compile itself is being done it will load the data through the inline function:
```php
$annotationCompiler->setAnnotationPatterns(function() { return JMSDiExtraBundle::getAnnotationPatterns(); });
```

I can change that to whatever you prefer :) just a further hint what you mean with the bundle configuration.

**Comment by Ener-Getick** [comment]
Hum this is why you allowed using a callback...
I'd prefer to add an ``arrayNode`` in [this file](https://github.com/schmittjoh/JMSDiExtraBundle/blob/master/DependencyInjection/Configuration.php)

**Comment by wodka** [comment]
but by adding it to the configuration a user would have always to setup ie SonataAdminBundle inside of the jms_di configuration block.

I think it would make sense to have both - an automatic registration of new paths as well as an easy option to setup those inside the config:

```yml
jms_di_extra:
  annotation_patterns:
    - "App\Annotation"
```
(always adding the jmsannotations)

**Comment by Ener-Getick** [comment]
We can make this the default value

**Comment by wodka** [comment]
I just added it to the configuration - can you check?

File: DependencyInjection/JMSDiExtraExtension.php [hunk 1]
```diff
@@ -54,6 +55,12 @@ public function load(array $configs, ContainerBuilder $container)
         $container->setParameter('jms_di_extra.disable_grep', $config['disable_grep']);
         $container->setParameter('jms_di_extra.doctrine_integration', $config['doctrine_integration']);
 
+        if (!empty($config['annotation_patterns']) && is_array($config['annotation_patterns'])) {
+            foreach ($config['annotation_patterns'] as $annotation_pattern) {
+                JMSDiExtraBundle::addAnnotationPattern($annotation_pattern);
```
Comment (Ener-Getick) [inline DependencyInjection/JMSDiExtraExtension.php]: Please use ``$container->setParameter()`` and ``$container->getParameter()`` instead. Third party bundles can change this bundle configuration to do what they want.

File: DependencyInjection/Compiler/AnnotationConfigurationPass.php [hunk 2]
```diff
@@ -18,32 +18,89 @@
...
 use Symfony\Component\DependencyInjection\ContainerBuilder;
 use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
 
+/**
+ * Search for annotation usage.
+ */
 class AnnotationConfigurationPass implements CompilerPassInterface
 {
+    /**
+     * @var string[]|callable
```
Comment (Ener-Getick) [inline DependencyInjection/Compiler/AnnotationConfigurationPass.php]: There is no need to support ``callable`` if we use the container parameters.

**Comment by wodka** [comment]
I changed it now to only use container parameters

**Comment by wodka** [comment]
I think this belongs here: https://github.com/schmittjoh/JMSDiExtraBundle/issues/239#issuecomment-184183092

there is also the question if the MetadataProcessorInterface is ok - it implies that the logic of how to handle the annotation works sits right inside the annotation itself.
(https://github.com/schmittjoh/JMSDiExtraBundle/pull/209)

File: DependencyInjection/Compiler/AnnotationConfigurationPass.php [hunk 3]
```diff
@@ -18,32 +18,75 @@
...
+        foreach ($this->getPatterns($container) as $pattern) {
+            $this->handlePattern($container, $directories, $pattern, $factory, $converter, $disableGrep);
+        }
+    }
+
+    /**
+     * @param ContainerBuilder $container
+     * @return \string[]
+     */
+    protected function getPatterns(ContainerBuilder $container)
```
Comment (Ener-Getick) [inline DependencyInjection/Compiler/AnnotationConfigurationPass.php]: Should be ``privateè`

File: DependencyInjection/JMSDiExtraExtension.php [hunk 4]
```diff
@@ -54,6 +55,19 @@ public function load(array $configs, ContainerBuilder $container)
         $container->setParameter('jms_di_extra.disable_grep', $config['disable_grep']);
         $container->setParameter('jms_di_extra.doctrine_integration', $config['doctrine_integration']);
 
+        if ($config['annotation_patterns']) {
+            if ($container->getParameterBag()->has('jms_di_extra.annotation_patterns')) {
```
Comment (Ener-Getick) [inline DependencyInjection/JMSDiExtraExtension.php]: I don't think that we should manage this case. People should use the bundle configuration.

File: Resources/doc/annotations.rst [hunk 5]
```diff
@@ -252,3 +252,76 @@ Automatically, registers the given class as a listener with the Doctrine ORM or
...
+    /**
+     * @Annotation
+     */
+    class Special implements MetadataProcessorInterface
+    {
+        public function processMetadata(ClassMetadata $metadata)
+        {}
+    }
+
+register your annotations, either inside your Bundle configuration or the app config,
```
Comment (Ener-Getick) [inline Resources/doc/annotations.rst]: Should be updated

File: Resources/doc/annotations.rst [hunk 6]
```diff
@@ -252,3 +252,76 @@ Automatically, registers the given class as a listener with the Doctrine ORM or
...
+            {
+                if ($container->getParameterBag()->has('jms_di_extra.annotation_patterns')) {
+                    $patterns = $container->getParameter('jms_di_extra.annotation_patterns');
+                } else {
+                    $patterns = array();
+                }
+
+                $patterns[] = 'AppBundle\Annotation';
+
+                $container->setParameter('jms_di_extra.annotation_patterns', $patterns);
```
Comment (Ener-Getick) [inline Resources/doc/annotations.rst]: This is a bad practice

File: DependencyInjection/Compiler/AnnotationConfigurationPass.php [hunk 7]
```diff
@@ -18,32 +18,75 @@
...
+    {
+        $patterns = array('JMS\DiExtraBundle\Annotation');
```
Comment (Ener-Getick) [inline DependencyInjection/Compiler/AnnotationConfigurationPass.php]: Would be better to store the container value and then remove it to not have to maintain BC for it.

File: Resources/doc/annotations.rst (hunk 6 above)
Comment (wodka) [inline Resources/doc/annotations.rst]: :/ now we have a loop

-> we shall not have a function to add new namespaces on the bundle and should handle extensions through container parameters
-> we shall not overwrite the container parameter.

(both solutions are not perfect, but I do not see another solution)

do you have a suggestion of how to add those patterns?

Comment (Ener-Getick) [inline Resources/doc/annotations.rst]: You should use overwrite the bundle configuration (see [this](http://symfony.com/doc/current/cookbook/bundles/prepend_extension.html))

Comment (wodka) [inline Resources/doc/annotations.rst]: or only have it in configuration? the more I think about it the better this solution becomes

Comment (wodka) [inline Resources/doc/annotations.rst]: thanks :) I will add this

Comment (Ener-Getick) [inline Resources/doc/annotations.rst]: What do you mean ?
I think we should only be able to define patterns in the configuration and BTW I think that ``JMS\DiExtraBundle\Annotation`` should be the default value but should not be always added.

File: DependencyInjection/Compiler/AnnotationConfigurationPass.php (hunk 7 above)
Comment (wodka) [inline DependencyInjection/Compiler/AnnotationConfigurationPass.php]: moved it to the configuration

File: Resources/doc/annotations.rst (hunk 6 above)
Comment (wodka) [inline Resources/doc/annotations.rst]: didn't see your comment for the PrependExtension. I just added it and updated the documentation (also the SonataAdmin Loading)

File: DependencyInjection/JMSDiExtraExtension.php (hunk 4 above)
Comment (wodka) [inline DependencyInjection/JMSDiExtraExtension.php]: it is now inside the bundle configuration :)

File: DependencyInjection/Compiler/AnnotationConfigurationPass.php (hunk 3 above)
Comment (wodka) [inline DependencyInjection/Compiler/AnnotationConfigurationPass.php]: done :)

File: DependencyInjection/Configuration.php [hunk 8]
```diff
@@ -73,6 +73,13 @@ public function getConfigTreeBuilder()
                             ->scalarNode('cache')->defaultValue('file')->cannotBeEmpty()->end()
                         ->end()
                     ->end()
+                    ->arrayNode('annotation_patterns')
+                        ->info('Allows you to configure additional annotation patterns. '
+                            .'Other bundles can automatically register patterns by calling JMSDiExtraBundle::addAnnotationPattern($pattern).')
```
Comment (Ener-Getick) [inline DependencyInjection/Configuration.php]: Should be removed.

File: DependencyInjection/Configuration.php [hunk 9]
```diff
@@ -73,6 +73,13 @@ public function getConfigTreeBuilder()
...
+                        ->treatNullLike(array('JMS\DiExtraBundle\Annotation'))
```
Comment (Ener-Getick) [inline DependencyInjection/Configuration.php]: Use ``addDefaultsIfNotSet`` instead.

File: DependencyInjection/Compiler/AnnotationConfigurationPass.php [hunk 10]
```diff
@@ -18,32 +18,66 @@
...
+        }
+    }
+
+    /**
+     * @param ContainerBuilder $container
+     * @return \string[]
+     */
+    private function getPatterns(ContainerBuilder $container)
+    {
+        return $container->getParameter('jms_di_extra.annotation_patterns');
```
Comment (Ener-Getick) [inline DependencyInjection/Compiler/AnnotationConfigurationPass.php]: please replace this by something like
```php
private $patterns;

private function getPatterns(ContainerBuilder $container)
{
     if (null === $this->patterns) {
         $this->patterns = $container->getParameter('jms_di_extra.annotation_patterns');
         $container->removeParameter('jms_di_extra.annotation_patterns');
     }

     return $this->patterns;
}

File: DependencyInjection/Configuration.php (hunk 8 above)
Comment (wodka) [inline DependencyInjection/Configuration.php]: done

File: Resources/doc/annotations.rst [hunk 11]
```diff
@@ -252,3 +252,85 @@ Automatically, registers the given class as a listener with the Doctrine ORM or
...
+~~~~~~~~~~~~~~~~~~
+Build your own annotations and register them
+
+.. code-block :: php
+
+    <?php
+
+    namespace AppBundle\Annotation;
+
+    use JMS\DiExtraBundle\Annotation\DoctrineListener;
```
Comment (Ener-Getick) [inline Resources/doc/annotations.rst]: That's not the good annotation here.

File: DependencyInjection/Configuration.php (hunk 9 above)
Comment (wodka) [inline DependencyInjection/Configuration.php]: done

File: Resources/doc/annotations.rst [hunk 12]
```diff
@@ -252,3 +252,85 @@ Automatically, registers the given class as a listener with the Doctrine ORM or
...
+    /**
+     * @Annotation
+     */
+    class Special implements MetadataProcessorInterface
+    {
+        public function processMetadata(ClassMetadata $metadata)
+        {}
+    }
+
+register your annotations, either as prepend extension or by adding it to your config.yml
```
Comment (Ener-Getick) [inline Resources/doc/annotations.rst]: I'm not sure that we should specify that people can prepend the bundle configuration in their extensions.
This is not what people should do on a project except if they are developing generic bundles.

Comment (Ener-Getick) [inline Resources/doc/annotations.rst]: BTW it looks strange to me to have 3 different namespaces in this section.

File: DependencyInjection/Compiler/AnnotationConfigurationPass.php (hunk 10 above)
Comment (wodka) [inline DependencyInjection/Compiler/AnnotationConfigurationPass.php]: I will add this without the removing of the parameter - cannot do that on a frozen parameter bag

Comment (Ener-Getick) [inline DependencyInjection/Compiler/AnnotationConfigurationPass.php]: It is not frozen, the container is built when this code is runned.

File: Resources/doc/annotations.rst (hunk 12 above)
Comment (wodka) [inline Resources/doc/annotations.rst]: I moved now all to a single namespace "App" and created a new section just for generic bundles

File: Resources/doc/annotations.rst (hunk 11 above)
Comment (wodka) [inline Resources/doc/annotations.rst]: removed

File: DependencyInjection/Configuration.php (hunk 9 above)
Comment (wodka) [inline DependencyInjection/Configuration.php]: ok, I cannot do it with addDefaultsIfNotSet

Symfony\Component\Config\Definition\Exception\InvalidDefinitionException: ->addDefaultsIfNotSet() is not applicable to prototype nodes at path "jms_di_extra.annotation_patterns"

Comment (Ener-Getick) [inline DependencyInjection/Configuration.php]: Oops right you should use ``addDefaultChildrenIfNoneSet``

File: DependencyInjection/Compiler/AnnotationConfigurationPass.php (hunk 10 above)
Comment (wodka) [inline DependencyInjection/Compiler/AnnotationConfigurationPass.php]: upps - added the remove now

File: DependencyInjection/Configuration.php [hunk 13]
```diff
@@ -73,6 +73,11 @@ public function getConfigTreeBuilder()
                             ->scalarNode('cache')->defaultValue('file')->cannotBeEmpty()->end()
                         ->end()
                     ->end()
+                    ->arrayNode('annotation_patterns')
+                        ->info('Allows you to configure additional annotation patterns.')
+                        ->addDefaultChildrenIfNoneSet()
```
Comment (Ener-Getick) [inline DependencyInjection/Configuration.php]: The default children should be added here
```php
->addDefaultChildrenIfNoneSet(array('JMS\DiExtraBundle\Annotation')

File: Resources/doc/annotations.rst [hunk 14]
```diff
@@ -252,3 +252,93 @@ Automatically, registers the given class as a listener with the Doctrine ORM or
...
+.. code-block :: php
+
+    <?php
+
+    namespace App;
+
+    use AppBundle\Annotation as App;
+
+    /**
+     * @Special(my_value="nice")
```
Comment (Ener-Getick) [inline Resources/doc/annotations.rst]: should be ``App\Special``

**Comment by Ener-Getick** [comment]
Merged in 9acab96d4942901cdc7c3564d23337269bd849e6.
Thank you @wodka

**Pull request closed without merging**